finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,719 @@
+    last_broadcast: HashMap<NodeIndex, Instant>,
+    last_response: HashMap<NodeIndex, Instant>,
+    requested_authentications: HashMap<NodeIndex, HashSet<NodeIndex>>,
+    // Used to rotate the nodes we query about unknown nodes.
+    next_query: NodeIndex,
+}
+
+type DiscoveryCommand = (DiscoveryMessage, DataCommand);
+
+fn authentication_broadcast(authentication: Authentication) -> DiscoveryCommand {
+    (
+        DiscoveryMessage::AuthenticationBroadcast(authentication),
+        DataCommand::Broadcast,
+    )
+}
+
+fn request(
+    missing_authorities: Vec<NodeIndex>,
+    authentication: Authentication,
+    peer_id: PeerId,
+) -> DiscoveryCommand {
+    (
+        DiscoveryMessage::Request(missing_authorities, authentication),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+fn response(authentications: Vec<Authentication>, peer_id: PeerId) -> DiscoveryCommand {
+    (
+        DiscoveryMessage::Authentications(authentications),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+impl Discovery {
+    /// Create a new discovery handler with the given response/broadcast cooldown.
+    pub fn new(cooldown: Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
+            next_query: NodeIndex(0),
+        }
+    }
+
+    fn should_broadcast(missing_authorities_num: usize, total_node_count: NodeCount) -> bool {
+        // If we are not sure we know of at least one honest node.
+        missing_authorities_num * 3 > 2 * total_node_count.0
+    }
+
+    /// Returns messages that should be sent as part of authority discovery at this moment.
+    pub fn discover_authorities(&mut self, handler: &SessionHandler) -> Vec<DiscoveryCommand> {
+        let missing_authorities = handler.missing_nodes();
+        if missing_authorities.is_empty() {
+            return Vec::new();
+        }
+        let node_count = handler.node_count();
+        let authentication = handler.authentication();
+        if Self::should_broadcast(missing_authorities.len(), node_count) {
+            // We know of fewer than 1/3 authorities, broadcast our authentication and hope others
+            // respond in kind.
+            vec![authentication_broadcast(authentication)]
+        } else {
+            // Attempt learning about more authorities from the ones you already know.
+            let mut result = Vec::new();
+            let mut target = self.next_query;
+            while result.len() < NODES_TO_QUERY {
+                if let Some(peer_id) = handler.peer_id(&target) {
+                    result.push(request(
+                        missing_authorities.clone(),
+                        authentication.clone(),
+                        peer_id,
+                    ));
+                }
+                target = NodeIndex((target.0 + 1) % node_count.0);
```

Comment:
Maybe also randomize?

Reply:
Same response as above, I believe this is good enough.